ledger/sandbox/src/main/scala/com/digitalasset/platform/apiserver/services/ApiSubmissionService.scala

```diff
@@ -99,10 +114,49 @@ final class ApiSubmissionService private (
     val failedInterpretationsMeter: Meter =
       metrics.meter("daml.lapi.command_submission_service.failed_command_interpretations")
 
+    val deduplicatedCommandsMeter: Meter =
+      metrics.meter("daml.lapi.command_submission_service.deduplicated_commands")
+
     val submittedTransactionsTimer: Timer =
       metrics.timer("daml.lapi.command_submission_service.submitted_transactions")
   }
 
+  private def deduplicateAndRecordOnLedger(commands: ApiCommands)(
+      implicit logCtx: LoggingContext): Future[Unit] = {
+    val deduplicationKey = commands.submitter + "%" + commands.commandId.unwrap
+    val submittedAt = Instant.now
+    val ttl = submittedAt.plusNanos(commands.ttl.getOrElse(configuration.maxTtl).toNanos)
+
+    submissionService.deduplicateCommand(deduplicationKey, submittedAt, ttl).flatMap {
+      case CommandDeduplicationNew =>
+        recordOnLedger(commands)
+          .andThen {
+            case Success(_) =>
+              submissionService.updateCommandResult(deduplicationKey, submittedAt, Right(()))
+            case Failure(error: StatusRuntimeException) =>
```

Comment:
If you only want to catch both `StatusRuntimeException`s and `StatusException`s you can use the extractors in `//libs-scala/grpc-utils`.